compiler/src/iree/compiler/DispatchCreation/PropagateEncodings.cpp

```diff
@@ -0,0 +1,122 @@
+
+#define DEBUG_TYPE "iree-dispatch-creation-propagate-encodings"
+
+namespace mlir::iree_compiler::DispatchCreation {
+
+#define GEN_PASS_DEF_PROPAGATEENCODINGSPASS
+#include "iree/compiler/DispatchCreation/Passes.h.inc"
+
+namespace {
+
+/// Pattern to swap `tensor.collapse_shape` -> `iree_encoding.set_encoding`
+struct SwapEncodingOpWithTensorCollapseShapeOp
+    : public OpRewritePattern<IREE::Encoding::SetEncodingOp> {
+  using Base = OpRewritePattern<IREE::Encoding::SetEncodingOp>;
+  using Base::Base;
+  LogicalResult matchAndRewrite(IREE::Encoding::SetEncodingOp encodingOp,
+                                PatternRewriter &rewriter) const override;
+};
+
+// TODO(#20179): Support the propagation through interfaces. It is supposed to
+// be done with data-flow analysis.
+struct PropagateEncodingsPass
+    : public DispatchCreation::impl::PropagateEncodingsPassBase<
+          PropagateEncodingsPass> {
+  void runOnOperation() override;
+};
+
+} // namespace
+
+LogicalResult SwapEncodingOpWithTensorCollapseShapeOp::matchAndRewrite(
+    IREE::Encoding::SetEncodingOp encodingOp, PatternRewriter &rewriter) const {
+  auto encoding = dyn_cast<IREE::Encoding::MatmulKAttr>(
+      encodingOp.getResultType().getEncoding());
+  if (!encoding) {
+    return rewriter.notifyMatchFailure(encodingOp, "only matmul_k is handled");
+  }
+  auto collapseOp =
+      encodingOp.getSource().getDefiningOp<tensor::CollapseShapeOp>();
+  if (!collapseOp) {
+    return rewriter.notifyMatchFailure(encodingOp,
+                                       "expected a collapse_shape producer");
+  }
+  if (!IREE::Flow::isNonNullAndOutsideDispatch(encodingOp) ||
+      !IREE::Flow::isNonNullAndOutsideDispatch(collapseOp)) {
+    return rewriter.notifyMatchFailure(
+        encodingOp, "expected that both operations are outside dispatch");
+  }
+
+  ArrayRef<int32_t> kDims = encoding.getKDims().asArrayRef();
+  llvm::SetVector<int32_t> kDimsSet(kDims.begin(), kDims.end());
+
+  // Bail out if it is not propagable.
+  // TODO: Relax the check to allow transforming innermost reduction dimensions.
+  // We need to revisit the matmul_k encoding semantic.
+  SmallVector<ReassociationIndices, 4> reassociationMaps =
+      collapseOp.getReassociationIndices();
+  for (int32_t k : kDims) {
+    if (reassociationMaps[k].size() != 1) {
+      return rewriter.notifyMatchFailure(
+          encodingOp,
+          "expected collaps_shape ops to not transform k dimensions");
+    }
+  }
+
+  // Get a mapping from original iteration space to expanded iteration space.
+  MLIRContext *ctx = rewriter.getContext();
+  unsigned numDims = 0;
+  SmallVector<int32_t> newKDims;
+  for (auto [idx, reassociation] : llvm::enumerate(reassociationMaps)) {
+    if (kDimsSet.count(idx)) {
+      newKDims.append(llvm::to_vector(
+          llvm::seq<int32_t>(numDims, numDims + reassociation.size())));
+    }
+    numDims += reassociation.size();
+  }
```

Comment:
nit: I don't think the numDims is necessary. It could just be:
```suggestion
  SmallVector<int32_t> newKDims;
  for (int32_t kDim : dDims) {
    newKDims.append(reassociationMaps[kDim]);
  }
```

Reply:
It could also be combined with the above verification loop if doing it this way.

Reply:
Good point, thanks!